### PR TITLE
[TASK] Better support for Dimensions

### PR DESCRIPTION
**Pull request by dfeyer**
This change take care of Node variants in all dimension combinations
during Indexing. This change introduce a new property in the Elasticsearch index called ```__dimensionCombinationHash``` so you need to reindex all your nodes.

### CODE REVIEW

**Comment by dfeyer** [comment]
I will test this on a big project this week-end please don't merge before my feedback

**Comment by dfeyer** [comment]
It doesn't work

```
"__dimensionCombinations.language": [
                  "fr_FR"
                ],
```

Match all dimension that contains "fr_FR", need to tune this query a bit

**Comment by fanat98** [comment]
For me works it +1

**Comment by dfeyer** [comment]
@fanat98 Can you post here your dimensions settings ?

For me, query in german work, but all query for french return french + german, with this configuration:
````
    contentDimensions:
      'language':
        default: 'fr_FR'
        defaultPreset: 'fr'
        label: 'Language'
        icon: 'icon-language'
        presets:
          'fr':
            label: 'French'
            values: ['fr_FR']
            uriSegment: 'fr'
          'de':
            label: 'German'
            values: ['de_DE', 'fr_FR']
            uriSegment: 'de'
```

I will push a new version of the PR with a new __dimensionHash property in ES index (md5 of the jsonify dimensionCombinations), easy to query and will only return nodes in the current dimensions.

**Comment by fanat98** [comment]
Hi @dfeyer here my dimensions settings


      contentDimensions:
        'language':
          icon: 'icon-language'
          default: 'de'
          defaultPreset: 'de'
          presets:
            'fr':
              label: 'Français'
              values: ['fr']
              uriSegment: 'fr'
            'de':
              label: 'Deutsch'
              values: ['de']
              uriSegment: 'de'
            'en': ~

**Comment by dfeyer** [comment]
Now it work with all type of dimensions configuration. During indexing, we take care to create one document per dimensions combination, and we store a hash of the combination. This hash is used in the query to filter correctly the result

**Breaking Change**: This change require a full reindexing to work

Maybe it's time to release the 2.0 ;)

:+1: from my side

**Comment by fanat98** [comment]
Awesome :+1:
Thank you @dfeyer !

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php [hunk 1]
```diff
@@ -773,8 +773,9 @@ public function query(NodeInterface $contextNode)
         $this->queryFilter('terms', array('__workspace' => array_unique(array('live', $contextNode->getContext()->getWorkspace()->getName()))));
 
         // match exact dimension values for each dimension, this works because the indexing flattens the node variants for all dimension preset combinations
-        foreach ($contextNode->getContext()->getDimensions() as $dimensionName => $dimensionValues) {
-            $this->queryFilter('terms', array('__dimensionCombinations.' . $dimensionName => $dimensionValues, 'execution' => 'and'));
+        $dimensionCombinations = $contextNode->getContext()->getDimensions();
+        if (is_array($dimensionCombinations)) {
+            $this->queryFilter('term', ['__dimensionHash' => md5(json_encode($dimensionCombinations))]);
```
Comment (dfeyer) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php]: Maybe ```__dimensionHash``` can be renamed ```__dimensionCombinationHash ``` to be more explicit

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Mapping/NodeTypeMappingBuilder.php [hunk 2]
```diff
@@ -109,6 +109,10 @@ public function buildMappingInformation(Index $index)
                     'index' => 'not_analyzed'
                 )
             ));
+            $mapping->setPropertyByPath('__dimensionHash', [
```
Comment (dfeyer) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Mapping/NodeTypeMappingBuilder.php]: Same here ```__dimensionCombinationHash``` can be more explicit

**Comment by daniellienert** [comment]
I fixed the tests.
:+1: by reading and testing

**Comment by daniellienert** [comment]
@dfeyer - there is already a 2.0.1 - but I think we could release a 2.1

**Comment by dfeyer** [comment]
Last version with squashed commit and a small change in ```ElasticSearchQueryTest::termSuggestion``` to test only the test content and not the ```freq``` property returned by ElasticSearch, because this value can change depending on the Elasticsearch version or the index content.

:+1: from my side and ready to merge

**Comment by skurfuerst** [comment]
:+1: by reading!

**Pull request merged**